src/DotNetty.Common/Utilities/StringUtil.cs

```diff
@@ -287,12 +299,196 @@ public static string EscapeCsv(string value)
+                            unescaped.Add((StringCharSequence)current.ToString());
+                            current.Length = 0;
+                            break;
+                        case DoubleQuote:
+                            if (current.Length == 0)
+                            {
+                                quoted = true;
+                            }
+                            else
+                            {
+                                // double-quote appears without being enclosed with double-quotes
+                                current.Append(c);
+                            }
+                            break;
+                        case LineFeed:
+                        case CarriageReturn:
+                            // special characters appears without being enclosed with double-quotes
+                            throw new ArgumentException($"invalid escaped CSV field: {value} index: {i}");
+                        default:
+                            current.Append(c);
+                            break;
+                    }
+                }
+            }
+            if (quoted)
+            {
+                throw new ArgumentException($"invalid escaped CSV field: {value} index: {last}");
+            }
+
+            unescaped.Add((StringCharSequence)current.ToString());
+            return unescaped;
+        }
+
+        public static int IndexOfNonWhiteSpace(IReadOnlyList<char> seq, int offset)
+        {
+            for (; offset < seq.Count; ++offset)
+            {
+                if (!char.IsWhiteSpace(seq[offset]))
+                {
+                    return offset;
+                }
+            }
+
+            return -1;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static bool IsSurrogate(char c) => c >= '\uD800' && c <= '\uDFFF';
+
+        public static bool EndsWith(IReadOnlyList<char> s, char c)
+        {
+            int len = s.Count;
+            return len > 0 && s[len - 1] == c;
+        }
+
+        public static ICharSequence TrimOws(ICharSequence value)
+        {
+            int length = value.Count;
+            if (length == 0)
+            {
+                return value;
+            }
+
+            int start = IndexOfFirstNonOwsChar(value, length);
+            int end = IndexOfLastNonOwsChar(value, start, length);
+            return start == 0 && end == length - 1 ? value : value.SubSequence(start, end + 1);
+        }
+
+        internal static int IndexOf(IReadOnlyList<char> value, char ch, int start)
+        {
+            Contract.Requires(start >= 0 && start < value.Count);
+
+            char upper = char.ToUpper(ch);
+            char lower = char.ToLower(ch);
+            int i = start;
+            while (i < value.Count)
+            {
+                char c1 = value[i];
+                if (c1 == ch
+                    && char.ToUpper(c1).CompareTo(upper) != 0
```

Comment:
Not sure I understand it but ascii string can do a byte compare which is much better.